pkg/logs/decoder/handler.go

```diff
@@ -35,12 +38,76 @@ func (s *SingleHandler) Handle(line *RichLine) {
 
 // Cleanup closes the downstream operations.
 func (s *SingleHandler) Cleanup() {
+	s.SendResult()
 	s.truncator.Close()
 }
 
 // SendResult does nothing since single line handler doesn't cache the lines.
 func (s *SingleHandler) SendResult() {}
 
+// MultiHandler handles multiline logs. It accumulates the multiline logs and
+// send it to downstream.
+type MultiHandler struct {
+	truncator   LineTruncator
+	newLogRegex *regexp.Regexp
+	buffer      *MultiLineBuffer
+}
+
+// NewMultiHandler creates a new instance of MultiHandler.
+func NewMultiHandler(newLogRegex *regexp.Regexp, truncator LineTruncator) *MultiHandler {
+	return &MultiHandler{
+		newLogRegex: newLogRegex,
+		truncator:   truncator,
+		buffer:      NewMultiLineBuffer(),
+	}
+}
+
+// Handle checks line content prefix against a regex to see if it's a start of new log.
+// It accumulates lines and merge as one log and then send to next handler.
+// If the specified line requires leading or tailing truncation information, treat
+// this line "specially" meaning sending them directly to down stream without buffering.
+func (m *MultiHandler) Handle(line *RichLine) {
+	if m.newLogRegex.Match(line.Content) {
+		// it's the start of a new log, handle buffered content.
+		m.SendResult()
+	}
+	m.cacheLine(line)
+	// When leading or tailing truncation info is required, the line should be
+	// part of large line splitted by upstream.
+	// in case of:
+	// msg...TRUNCATED... needTailing = true or
+	// ...TRUNCATED...msg...TRUNCATED... needLeading = true && needTailing = true
+	// both cases above suggest msg is at it's cap length, that no buffering is
+	// required.
+	if m.buffer.IsFull() {
+		m.SendResult()
+	}
+}
+
+// cacheLine appends the content of specified line to the end of buffer, if buffer was not empty,
+// a `\n` will be append prior to the content of this line.
+func (m *MultiHandler) cacheLine(line *RichLine) {
```

Comment:
we are not caching here, we are aggregating or buffering content: the purpose of caching is to prevent unnecessary I/O operation by storing results which can be further reused.